### PR TITLE
Add ViewChild and ViewChildren

### PR DESCRIPTION
**Pull request by mebibou**
This is a simplified version of the angular one, but I think this would be quite useful. I don't know if there is a way to check the elements still exist during the lifecycle of the component though

### CODE REVIEW

**Comment by vsternbach** [comment]
First of all thanks for contributing to the project!
Regarding you pull request, I am not sure if it really adds the same value as ViewChild decorator intends to, as per [documentation](https://angular.io/guide/component-interaction#parent-calls-an-viewchild) it basically gives you access from parent component's controller to child component's controller rather than giving reference to JQLite element of the parent's template based on the provided selector

**Comment by mebibou** [comment]
I know the idea should be to use `@ViewChild(ChildComponent)` but I'm not sure how to do this here

File: src/query.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * @stable
+ */
+export interface IQueries {
+  [property: string]: {
+    first: boolean;
+    selector: string;
+  }
+}
+
+export function ViewChild(selector: string): any {
```
Comment (vsternbach) [inline src/query.ts]: change to (type: any), same for the ViewChildren decorator

File: src/query.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  return (target: any, key: string) => addBindingToMetadata(target, key, selector, true);
+}
+
+export function ViewChildren(selector: string): any {
+  return (target: any, key: string) => addBindingToMetadata(target, key, selector, false);
+}
+
+function addBindingToMetadata(target: any, key: string, selector: string, first: boolean) {
```
Comment (vsternbach) [inline src/query.ts]: also change selector to type

File: src/query.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  const targetConstructor = target.constructor;
+  const queries: IQueries = getMetadata(metadataKeys.queries, targetConstructor) || {};
+  queries[key] = {
```
Comment (vsternbach) [inline src/query.ts]: const selector = camelToKebab(getTypeName(type));
both methods are in utils.ts

File: src/query.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { defineMetadata, getMetadata, metadataKeys } from './utils';
+import { Type } from './type';
+
+/**
+ * Type of the Query metadata.
+ *
+ * @stable
+ */
+export interface IQueries {
```
Comment (vsternbach) [inline src/query.ts]: I would change the name to IViewChildren and the file name to view_child

File: src/utils.ts [hunk 5]
```diff
@@ -10,6 +10,7 @@ export const metadataKeys = {
   bindings: 'custom:bindings',
   options: 'custom:options',
   listeners: 'custom:listeners',
+  queries: 'custom:queries',
```
Comment (vsternbach) [inline src/utils.ts]: would change also to viewChildren

File: src/component.ts [hunk 6]
```diff
@@ -68,12 +72,24 @@ export function extendWithHostListeners(ctrl: {new(...args: any[])}, listeners:
         const { eventName } = listeners[handler];
         this.$element.on(eventName + namespace, this[handler].bind(this));
       });
+      properties.forEach(property => {
+        const query = queries[property];
+        const elements = this.$element[0].querySelector(query.selector);
+        if (query.first) {
+          this[property] = angular.element(elements[0]);
```
Comment (vsternbach) [inline src/component.ts]:
```
const viewChildEl = this.$element[0].querySelector(query.selector);
this[property] = angular.element(viewChildEl).isolateScope<any>()['$ctrl'];
```

File: src/component.ts [hunk 7]
```diff
@@ -68,12 +72,24 @@ export function extendWithHostListeners(ctrl: {new(...args: any[])}, listeners:
...
+        }
+        else {
+          this[property] = angular.element(elements);
```
Comment (vsternbach) [inline src/component.ts]:
```
const viewChildEls = this.$element[0].querySelectorAll(query.selector);
this[property] = viewChildEls.map(viewChild => angular.element(viewChild).isolateScope<any>()['$ctrl'])
```

Comment (vsternbach) [reply src/component.ts]: It's possible to pass child component's type to the decorator and get it's selector from it's name when you define view children , and then you can get to $ctrl property on isolateScope() method of angular.element, that is viewChild controller.
Left you comments in code that need to be changed.
In general I would rename all the "query" names to "viewChild"

**Comment by vsternbach** [comment]
It's possible to pass child component's type to the decorator and get it's selector from it's name when you define view children , and then you can get to $ctrl property on isolateScope() method of angular.element, that is viewChild controller.
Left you comments in code that need to be changed.
In general I would rename all the "query" names to "viewChild"

**Comment by mebibou** [comment]
I've made the changes according to your feedback. I don't think I am testing anything though in the tests, and also shouldn't the selector be `Type<any>|string` ?

**Comment by vsternbach** [comment]
You 're probably right, the interface of the decorator should be like this:
(selector: Type<any>|Function|string, opts?: {read?: any}): any;
But I need to figure out first what's the usecase for using opts.read and how to use it
I'll merge the changes anyway for now and add my own commit
Thanks

**Comment by mebibou** [comment]
Actually just realized you cannot do `.map` on the result of `. querySelectorAll` (I'm using a custom function in the meantime before you release this new feature), but the theory works well ;)

**Comment by vsternbach** [comment]
Why not, the result of querySelectAll is not iterable?

**Comment by mebibou** [comment]
seems like it, Chrome was throwing error `querySelectAll(...).map` is not a function, since it returns a [NodeList](https://developer.mozilla.org/fr/docs/Web/API/NodeList)

**Comment by vsternbach** [comment]
you can use destructuring for this: [...document.body.querySelectorAll('div')].map(node => node)
I'll change it, thanks

**Pull request merged**